Common++/header/PointerVector.h

```diff
@@ -55,10 +55,29 @@ namespace pcpp
 		 */
 		PointerVector(const PointerVector& other)
 		{
-			for (const auto iter : other)
+			try
 			{
-				T* objCopy = new T(*iter);
-				m_Vector.push_back(objCopy);
+				for (const auto iter : other)
+				{
+					T* objCopy = new T(*iter);
+					try
+					{
+						m_Vector.push_back(objCopy);
+					}
+					catch (const std::exception&)
+					{
+						delete objCopy;
+						throw;
+					}
+				}
+			}
+			catch (const std::exception&)
+			{
+				for (auto obj : m_Vector)
```

Comment:
this part of logic needs to be fixed, you just delete the pointer of the element in the vector, but you didn't remove the element.

Reply:
correct me if i am wrong,  I think the correct logic here should be
 ```cpp
 for (const auto iter : other)
{
    try 
    {
        T* objCopy = new T(*iter);
    } catch (std::exception& e) {
        delete objCopy;
        throw e;
    }
    try{
        m_Vector.push_back(objCopy);
    } catch (std::exception& e) {
        throw e;
    }
}
```

Reply:
Uh, not really. This has the exact same issue from where we started, but with extra steps.

If the new statement or the ctor for T in it throws, the memry allocated by new is automatically freed. Also objectCopy isnt really assigned so you don't have anything to delete.

`push_back` has a strong exception guarantee in the standard. Which means if the push failed, the vector is rolled back to the last valid state (pre-push), so the pointer is not added. That is at the core of the whole memory leak issue. If the push fails, the ptr is not added and you lose the `objCpy`, so you have to `delete` the memory as the exception is being propagated outwards before you leave the function scope.

Reply:
I understand now, thank you for clarifying. As per the current code logic, if an element cannot be successfully pushed into `m_Vector`, it triggers an exception within the inner try block. The outer try then catches this exception and enters the corresponding catch block, leading to the deletion of all pointers in `m_Vector`. Nevertheless, following these steps, the elements within `m_Vector` persist, but they now reference invalid memory, posing a potential issue?

Reply:
Since this is a copy constructor the rules for throwing exceptions out of ctors apply. All initialized member fields have their destructors called as the exception propagates out of the ctor and the object destructor is not called as it did not finish construction. 

Therefor `m_Vector` is deallocated wholesale as the exception leaves the ctor, so it is pointless to clear the vector if we don't have any complex logic using it. As it is, the exception is immediately rethrown so we skip to member field deallocation after the outer try block.

Reply:
True, thanks for correcting me.